test/metabase/models/card_test.clj

```diff
@@ -210,3 +210,58 @@
+
+;; this is a separate function so we can use the same tests for DashboardCards as well
+(defn test-visualization-settings-normalization [f]
+  (testing "visualization settings should get normalized to use modern MBQL syntax"
+    (testing "Field references in column settings"
+      (doseq [[original expected] {[:ref [:field-literal "foo" :type/Float]]
+                                   [:ref [:field "foo" {:base-type :type/Float}]]
+
+                                   [:ref [:field-id 1]]
+                                   [:ref [:field 1 nil]]
+
+                                   [:ref [:expression "wow"]]
+                                   [:ref [:expression "wow"]]}
+              ;; also check that normalization of already-normalized refs is idempotent
+              original [original expected]
+              ;; frontend uses JSON-serialized versions of the MBQL clauses as keys
+              :let     [original (json/generate-string original)
+                        expected (json/generate-string expected)]]
+        (testing (format "Viz settings field ref key %s should get normalized to %s"
+                         (pr-str original)
+                         (pr-str expected))
+          (f
+           {:column_settings {original {:currency "BTC"}}}
+           {:column_settings {expected {:currency "BTC"}}}))))
+
+    (testing "Other MBQL field clauses"
+      (let [original {:map.type                 "region"
+                      :map.region               "us_states"
+                      :pivot_table.column_split {:rows    [["datetime-field" ["field-id" 807] "year"]],
+                                                 :columns [["fk->" ["field-id" 805] ["field-id" 808]]],
+                                                 :values  [["aggregation" 0]]}}
+            expected {:map.type                 "region"
+                      :map.region               "us_states"
+                      :pivot_table.column_split {:rows    [[:field 807 {:temporal-unit :year}]]
+                                                 :columns [[:field 808 {:source-field 805}]]
+                                                 :values  [[:aggregation 0]]}}]
+        (f original expected)))
+
+    (testing "Don't normalize non-MBQL arrays"
+      (let [original {:graph.show_goal  true
+                      :graph.goal_value 5.9
+                      :graph.dimensions ["the_day"]
+                      :graph.metrics    ["total_per_day"]}]
+        (f original original)))
+
+    (testing "Don't normalize key-value pairs in maps that could be interpreted as MBQL clauses"
+      (let [original {:field-id 1}]
+        (f original original)))))
+
+(deftest normalize-visualization-settings-test
+  (test-visualization-settings-normalization
+   (fn [original expected]
+     (mt/with-temp Card [card {:visualization_settings original}]
+       (is (= expected
+              (db/select-one-field :visualization_settings Card :id (u/the-id card))))))))
```

Comment:
@paulrosenzweig I added tests for normalization of serialized Card/DashboardCard field references as keys in viz settings here (specifically the `json/generate-string` part at the top). Are the serialized keys used anywhere else besides `visualization_settings.column_settings`?

Reply:
> Are the serialized keys used anywhere else besides visualization_settings.column_settings?

Sorry slow responding to that. I had thought both conditional formatting and pivot tables did, but I just checked and neither do! So thankfully I guess it's just column settings.